R/compat-plyr.R

```diff
@@ -359,19 +359,43 @@ rbind_dfs <- function(dfs) {
 #' @noRd
 dapply <- function(df, by, fun, ..., drop = TRUE) {
   grouping_cols <- .subset(df, by)
-  ids <- id(grouping_cols, drop = drop)
-  group_rows <- split(seq_len(nrow(df)), ids)
   fallback_order <- unique(c(by, names(df)))
-  rbind_dfs(lapply(seq_along(group_rows), function(i) {
-    cur_data <- df_rows(df, group_rows[[i]])
-    res <- fun(cur_data, ...)
+  apply_fun <- function(x) {
+    res <- fun(x, ...)
     if (is.null(res)) return(res)
     if (length(res) == 0) return(new_data_frame())
-    vars <- lapply(setNames(by, by), function(col) .subset2(cur_data, col)[1])
+    vars <- lapply(setNames(by, by), function(col) .subset2(x, col)[1])
     if (is.matrix(res)) res <- split_matrix(res)
     if (is.null(names(res))) names(res) <- paste0("V", seq_along(res))
     if (all(by %in% names(res))) return(new_data_frame(unclass(res)))
     res <- modify_list(unclass(vars), unclass(res))
     new_data_frame(res[intersect(c(fallback_order, names(res)), names(res))])
+  }
+
+  # Shortcut when only one group
+  if (all(vapply(grouping_cols, single_value, logical(1)))) {
+    return(apply_fun(df))
+  }
+
+  ids <- id(grouping_cols, drop = drop)
```

Comment:
This is likely to be significantly faster if we use the vctrs equivalent.

Reply:
Yes... I think most of the compatibility-plyr file will be erased when we decide to go full on with vctrs